packages/block-editor/src/components/block-edit/edit.js

```diff
@@ -48,27 +57,225 @@ const Edit = ( props ) => {
 	const context = useMemo( () => {
-		return blockType && blockType.usesContext
-			? Object.fromEntries(
-					Object.entries( blockContext ).filter( ( [ key ] ) =>
-						blockType.usesContext.includes( key )
-					)
-			  )
-			: DEFAULT_BLOCK_CONTEXT;
-	}, [ blockType, blockContext ] );
+		const _context =
+			blockType && blockType.usesContext
+				? Object.fromEntries(
+						Object.entries( blockContext ).filter( ( [ key ] ) =>
+							blockType.usesContext.includes( key )
+						)
+				  )
+				: DEFAULT_BLOCK_CONTEXT;
+		// Add block bindings context.
+		const blockBindingsContext = {};
+		if ( attributes?.metadata?.bindings ) {
+			Object.values( attributes?.metadata?.bindings || {} ).forEach(
+				( binding ) => {
+					registeredSources[ binding?.source ]?.usesContext?.forEach(
+						( key ) => {
+							blockBindingsContext[ key ] = blockContext[ key ];
+						}
+					);
+				}
+			);
+		}
+		return { ..._context, ...blockBindingsContext };
+	}, [
+		blockType,
+		blockContext,
+		attributes?.metadata?.bindings,
+		registeredSources,
+	] );
+
+	const computedAttributes = useSelect(
+		( select ) => {
+			if ( ! blockBindings ) {
+				return attributes;
+			}
+
+			const attributesFromSources = {};
+			const blockBindingsBySource = new Map();
+
+			for ( const [ attributeName, binding ] of Object.entries(
+				blockBindings
+			) ) {
+				const { source: sourceName, args: sourceArgs } = binding;
+				const source = registeredSources[ sourceName ];
+				if ( ! source || ! canBindAttribute( name, attributeName ) ) {
+					continue;
+				}
+
+				blockBindingsBySource.set( source, {
+					...blockBindingsBySource.get( source ),
+					[ attributeName ]: {
+						args: sourceArgs,
+					},
+				} );
+			}
+
+			if ( blockBindingsBySource.size ) {
+				for ( const [ source, bindings ] of blockBindingsBySource ) {
+					// Get values in batch if the source supports it.
+					let values = {};
+					if ( ! source.getValues ) {
+						Object.keys( bindings ).forEach( ( attr ) => {
+							// Default to the the source label when `getValues` doesn't exist.
+							values[ attr ] = source.label;
+						} );
+					} else {
+						values = source.getValues( {
+							select,
+							context,
+							clientId,
+							bindings,
+						} );
+					}
+					for ( const [ attributeName, value ] of Object.entries(
+						values
+					) ) {
+						if (
+							attributeName === 'url' &&
+							( ! value || ! isURLLike( value ) )
+						) {
+							// Return null if value is not a valid URL.
+							attributesFromSources[ attributeName ] = null;
+						} else {
+							attributesFromSources[ attributeName ] = value;
+						}
+					}
+				}
+			}
+
+			return {
+				...attributes,
+				...attributesFromSources,
+			};
+		},
+		[
+			attributes,
+			blockBindings,
+			clientId,
+			context,
+			name,
+			registeredSources,
+		]
+	);
+
+	const hasPatternOverrideDefault = hasPatternOverridesDefaultBindings(
```

Comment:
`hasPatternOverridesDefault` would align better. Overall, I don't like the current name, but it's used in two places and needs to be set as a dependency of `useCallback`, so it's better to go with a boolean value.